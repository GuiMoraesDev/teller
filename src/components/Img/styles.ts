import styled from 'styled-components';

import { StyledProps } from '.';

export const ImageWrapper = styled.span`
	height: 100%;

	border: none;
`;

export const Image = styled.img<StyledProps>`
	max-width: 100%;
	max-height: 100%;

	border-radius: ${({ theme }) => theme.rounded.full};

	border: ${({ theme, isBordered }) =>
		isBordered ? `3px solid ${theme.themeColors.canvas}` : 'none'};
`;
