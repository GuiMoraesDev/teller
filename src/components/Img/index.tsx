import { ImgHTMLAttributes, useState, useCallback } from 'react';

import chroma from 'chroma-js';

import * as Styles from './styles';

export interface StyledProps {
	isBordered?: boolean;
}

interface Props extends StyledProps, ImgHTMLAttributes<HTMLImageElement> {
	src?: string;
	alt?: string;
}

const Img = ({
	src,
	alt = 'Not found',
	isBordered = false,
	...props
}: Props): JSX.Element => {
	const [imageSrc, setImageSrc] = useState<string>(src || '');

	const handleErrorImages = useCallback(() => {
		const canvasElement = document.createElement('canvas');
		const context = canvasElement.getContext('2d') as CanvasRenderingContext2D;

		canvasElement.width = 220;
		canvasElement.height = 220;

		context.fillStyle = String(chroma.random().darken(2));
		context.fillRect(0, 0, canvasElement.width, canvasElement.height);

		context.font = '100px sans-serif';
		context.fillStyle = '#FFF';
		context.textAlign = 'center';
		context.textBaseline = 'middle';

		const textSplitted = alt.split(' ');
		const textFirstLetters = textSplitted.map((text) => text.charAt(0));
		const textCleanEmpty = textFirstLetters.filter((text) => !!text);
		const textJoined = textCleanEmpty.join('');

		context.fillText(
			`${textJoined.slice(0, 2).toUpperCase()}`,
			canvasElement.width / 2,
			canvasElement.height / 2
		);

		const dataUrl = canvasElement.toDataURL();

		setImageSrc(dataUrl);
	}, [alt]);

	return (
		<Styles.ImageWrapper>
			<Styles.Image
				{...props}
				src={imageSrc}
				alt={alt}
				onError={handleErrorImages}
				isBordered={isBordered}
				referrerPolicy="no-referrer"
				loading="lazy"
			/>
		</Styles.ImageWrapper>
	);
};

export default Img;
