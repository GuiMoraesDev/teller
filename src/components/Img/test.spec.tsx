import React from 'react';

import { render } from 'tests/test.utils';

import Img from './index';

describe('Img component', () => {
	it('should render correctly without isBordered prop', () => {
		const { container } = render(
			<Img src="https://i.pravatar.cc/150?id=50" alt="avatar placeholder" />
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly with isBordered prop', () => {
		const { container } = render(
			<Img
				src="https://i.pravatar.cc/150?id=50"
				alt="avatar placeholder"
				isBordered
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});
});
