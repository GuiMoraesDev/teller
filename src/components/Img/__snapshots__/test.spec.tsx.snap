// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Img component should render correctly with isBordered prop 1`] = `
.c0 {
  height: 100%;
  border: none;
}

.c1 {
  max-width: 100%;
  max-height: 100%;
  border-radius: 10000px;
  border: 3px solid #FDFDFD;
}

<span
  class="c0"
>
  <img
    alt="avatar placeholder"
    class="c1"
    loading="lazy"
    referrerpolicy="no-referrer"
    src="https://i.pravatar.cc/150?id=50"
  />
</span>
`;

exports[`Img component should render correctly without isBordered prop 1`] = `
.c0 {
  height: 100%;
  border: none;
}

.c1 {
  max-width: 100%;
  max-height: 100%;
  border-radius: 10000px;
  border: none;
}

<span
  class="c0"
>
  <img
    alt="avatar placeholder"
    class="c1"
    loading="lazy"
    referrerpolicy="no-referrer"
    src="https://i.pravatar.cc/150?id=50"
  />
</span>
`;
